File: Cracking the Coding Interview/Linked Lists- Detect a Cycle/LinkedListsDetectaCycle.py
"""
Detect a cycle in a linked list. Note that the head pointer may be 'None' if the list is empty.

A Node is defined as: 
 
    class Node(object):
        def __init__(self, data = None, next_node = None):
            self.data = data
            self.next = next_node
"""


def has_cycle(head):
    if head == None:
        return 0
    
    traversed = [0]*100;
    counter = 1
    nextOne = head.next
    
    while True:
        if nextOne == None:
            return 0
        if nextOne is head:
            return 1 - min(1, sum(traversed))
        traversed[counter] += 1
        counter += 1
        nextOne = nextOne.next

File: Cracking the Coding Interview/Linked Lists- Detect a Cycle/LinkedListsDetectaCycle.java
// https://www.hackerrank.com/challenges/ctci-linked-list-cycle/problem

/*
Detect a cycle in a linked list. Note that the head pointer may be 'null' if the list is empty.

A Node is defined as: 
    class Node {
        int data;
        Node next;
    }
*/

boolean hasCycle(Node head) {
    Node fast = head;
    while (head != null && fast != null && fast.next != null) {
        fast = fast.next.next;
        head = head.next;
        if (fast != null && head != null && fast.equals(head)) {
            return true;
        }
    }
    return false;
}
